### PR TITLE
Filter kafka cluster by topic

### PR DESCRIPTION
**Pull request by Will-Lo**
Addresses issue #697 

Adds topic parameter in `GET /kafka_cluster_state` to filter verbose output based on topic parameter with matching regex

### CODE REVIEW

**Comment by kidkun** [comment]
@Will-Lo Thanks for the patch!

@efeg I feel there are some confusions in the original description in issue #697.
Looking at the following verbose response from `kafka_cluster_state` endpoint, do we want to filter the replica list only in `Other Partition` field based on this `topic` regex or we want to filter data in all fields( include `Brokers`, `Offline Partitions` etc.)? I feel previous design makes more sense.

image

**Comment by efeg** [comment]
@kidkun @Will-Lo 
I was leaning towards applying the filter to _all replicas_ regardless of their current health status. In other words, the filter will be applied to each of the individual set (if relevant):

* `Offline Partitions`
* `Partitions with Offline Replicas` (relevant to `migrate_to_kafka_2_0` branch)
* `Under Replicated Partitions`
* `Under MinIsr Partitions` (relevant to `migrate_to_kafka_2_0` branch)
* `Other Partitions`

-- @Will-Lo Note that you can ignore the sets relevant to `migrate_to_kafka_2_0` branch -- we will apply the relevant changes while cherry-picking the patch.

**Comment by Will-Lo** [comment]
@efeg I changed the PR to only filter the topics on the partitions you mentioned by only populating the partitions that match the topic on `populateKafkaPartitionState`, which should cover underreplicated, offline, and other partitions. Is that okay?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/KafkaClusterState.java [hunk 1]
```diff
@@ -68,19 +71,22 @@ private String getJSONString(CruiseControlParameters parameters) {
   private void populateKafkaPartitionState(Set<PartitionInfo> underReplicatedPartitions,
                                            Set<PartitionInfo> offlinePartitions,
                                            Set<PartitionInfo> otherPartitions,
-                                           boolean verbose) {
+                                           boolean verbose,
+                                           Pattern topicPattern) {
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/KafkaClusterState.java]: Can we also update the Javadoc for this new parameter?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/KafkaClusterState.java [hunk 2]
```diff
@@ -155,7 +161,7 @@ private void populateKafkaBrokerState(Map<Integer, Integer> leaderCountByBrokerI
    *
    * @param verbose True if verbose, false otherwise.
    */
-  public Map<String, Object> getJsonStructure(boolean verbose) {
+  public Map<String, Object> getJsonStructure(boolean verbose, Pattern topic) {
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/KafkaClusterState.java]: Can we also update the Javadoc for this new parameter?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/KafkaClusterState.java [hunk 3]
```diff
@@ -52,7 +53,9 @@ public KafkaClusterState(Cluster kafkaCluster, KafkaCruiseControlConfig config)
 
   private String getJSONString(CruiseControlParameters parameters) {
     Gson gson = new Gson();
-    Map<String, Object> jsonStructure = getJsonStructure(((KafkaClusterStateParameters) parameters).isVerbose());
+    boolean isVerbose = ((KafkaClusterStateParameters) parameters).isVerbose();
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/KafkaClusterState.java]: nit: in this method and in `getPlaintext` we do cast form `CruiseControlParameters` to `KafkaClusterStateParameters` couple of times, maybe we can save one redundant cast.

**Comment by Will-Lo** [comment]
@kidkun Thanks for the feedback! I addressed the PR accordingly.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/KafkaClusterStateParameters.java [hunk 4]
```diff
@@ -14,11 +15,12 @@
  *
  * <pre>
  * Retrieve the kafka cluster state.
- *    GET /kafkacruisecontrol/kafka_cluster_state?verbose=[true/false]&amp;json=[true/false]
+ *    GET /kafkacruisecontrol/kafka_cluster_state?verbose=[true/false]&amp;json=[true/false];topic=[topic]
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/KafkaClusterStateParameters.java]: Nit: `;topic=[topic]` -> `&amp;topic=[topic]`

**Pull request merged**